### PR TITLE
rewrite json to msgpack and back and remove 3rdparty dependency

### PR DESCRIPTION
**Pull request by hendrikmuhs**
rewrite of how json is parse to msgpack and back, removing the 3rdparty dependency and replacing it with a simpler implementation

Notes:

- encoding is slightly faster (micro-benchmark)
- decoding avoids the temporary conversion to rapidjson which makes it faster, too
- the removal of the xchange dependency makes it possible to update msgpack-c (tbd in a separate PR)

### CODE REVIEW

**Comment by hendrikmuhs** [comment]
the only build failure left is the rust build, which is currently broken for master

**Comment by narekgharibyan** [comment]
@hendrikmuhs I'll look into rust build failure.

File: keyvi/include/keyvi/util/msgpack_util.h [hunk 1]
```diff
@@ -38,69 +40,111 @@
...
+      writer->Uint64(o.via.u64);
+      break;
+
+    case msgpack::type::NEGATIVE_INTEGER:
+      writer->Int64(o.via.i64);
+      break;
+
+    case msgpack::type::FLOAT:
+      writer->Double(o.via.f64);
+      break;
```
Comment (narekgharibyan) [inline keyvi/include/keyvi/util/msgpack_util.h:109]: In [JsonToMsgPack](https://github.com/KeyviDev/keyvi/pull/172/files#diff-9b644e267a5fc2eb412c40993bfd5740eb1cc692ca8801a1be8c5209dc15fc02R61-R66) we are writing floating point numbers calling 2 different methods: `pack_float()` and `pack_double()` and they have different type codes in `msgpack` format spec.

But in here while reading we have case just for `msgpack::type::FLOAT`, not sure if this is enough or we should have separate case for `msgpack::type::DOUBLE` as well.

File: keyvi/tests/keyvi/util/json_value_test.cpp [hunk 2]
```diff
@@ -23,22 +23,25 @@
...
+  std::string input =
+      "{\"hello\":\"world\",\"t\":true,\"f\":false,\"n\":null,\"i\":123,\"j\":-123,\"pi\":3.1416,\"a\":[1,2,3,4],\"d\":"
+      "{\"k\":\"v\"}}";
+
   std::string encoded = EncodeJsonValue(input);
 
   std::string output = DecodeJsonValue(encoded);
 
-  BOOST_CHECK_EQUAL('"' + input + '"', output);
+  BOOST_CHECK_EQUAL(input, output);
```
Comment (narekgharibyan) [inline keyvi/tests/keyvi/util/json_value_test.cpp:44]: will be good to add a test case with `single_precision_float=true`, as we have a special handling for that.

**Review by narekgharibyan — APPROVED** [review]
LGTM!

Left a few suggestions/question.

File: keyvi/tests/keyvi/util/json_value_test.cpp (hunk 2 above)
Comment (hendrikmuhs) [reply keyvi/tests/keyvi/util/json_value_test.cpp:44]: :+1: I assumed we have this covered, but it seems there is only a very basic test here:

https://github.com/KeyviDev/keyvi/blob/master/python/tests/json/json_dictionary_test.py#L59

I will add some coverage on C++ side.

File: keyvi/include/keyvi/util/msgpack_util.h (hunk 1 above)
Comment (hendrikmuhs) [reply keyvi/include/keyvi/util/msgpack_util.h:109]: `msgpack::type::DOUBLE` is deprecated, we do not use it (and if, it's an alias for `FLOAT`). Encoding and decoding is not 1:1. That means when reading back there is no difference and json does not care either.

Comment (hendrikmuhs) [reply keyvi/include/keyvi/util/msgpack_util.h:109]: FWIW I think I saw a distinction in newer versions for msgpack-c, will keep that in mind, once I update it, which is what I plan for after merging this PR.

Comment (narekgharibyan) [reply keyvi/include/keyvi/util/msgpack_util.h:109]: Cool, btw master is green now.

**Pull request merged**